go/libraries/doltcore/schema/index_coll.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema
+
+import "fmt"
+
+//TODO: column renames need to change colToTag
+//TODO: tag changes need to do...a lot
+type IndexCollection interface {
+	// AddIndexCol adds an index with the given name, columns (in index order), and table primary keys (also in order).
+	// Ex. For table {k1, k2, k3} PK(k1, k2) IDX(k3, k2), columns should be (k3, k2) and pks (k1, k2)
+	AddIndexCol(indexName string, cols []string, pks []string) (Index, error)
+	// AddIndexTag adds an index with the given name, column allTags (in index order), and table pk allTags (also in order).
+	AddIndexTag(indexName string, tags []uint64, pks []uint64) (Index, error)
+	// AllIndexes returns a slice containing all of the indexes in this collection.
+	AllIndexes() []Index
+	// BestMatch takes a collection of columns and finds the index that best matches, if one exists.
+	BestMatch(cols ...string) Index
+	// Contains returns whether the given index name already exists for this table.
+	Contains(indexName string) bool
+	// ContainsColumnCollection returns whether the collection contains an index that has this exact collection and ordering of columns.
+	ContainsColumnCollection(cols ...string) bool
+	// ContainsColumnTagCollection returns whether the collection contains an index that has this exact collection and ordering of columns.
+	ContainsColumnTagCollection(tags ...uint64) bool
+	// Count returns the number of indexes in this collection.
+	Count() int
+	// Get returns the index with the given name, or nil if it does not exist.
+	Get(indexName string) Index
+	// DropIndex removes an index from the table metadata.
+	DropIndex(indexName string) (Index, error)
+	// HasIndexes returns whether this collection has any indexes.
+	HasIndexes() bool
+	// RenameIndex renames an index in the table metadata.
+	RenameIndex(oldName, newName string) (Index, error)
+}
+
+type indexCollectionImpl struct {
+	indexes       map[string]*indexImpl
+	colTagToIndex map[uint64][]*indexImpl
+	colToTag      map[string]uint64
+}
+
+func NewIndexCollection(cols *ColCollection) IndexCollection {
```

Comment:
If the add methods require a schema obj, then you don't need a col collection to create a new one of these